### PR TITLE
Feature/overdue debts

### PR DESCRIPTION
**Pull request by NinoRy**
#

### CODE REVIEW

File: modules/BillingBase/Http/Controllers/SettlementRunController.php [hunk 1]
```diff
@@ -183,7 +183,7 @@ public function check_state()
         $response = new \Symfony\Component\HttpFoundation\StreamedResponse(function () {
             $job = true;
             while ($job) {
-                $job = \DB::table('jobs')->find(\Session::get('SrJobId'));
+                $job = \DB::table('jobs')->find(\Session::get('srJobId'));
```
Comment (cschra) [inline modules/BillingBase/Http/Controllers/SettlementRunController.php]: I would recommend to "use" DB, Session and Log as it is used quiet often

File: modules/OverdueDebts/Console/ParseMt940.php [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+    /**
+     * Get the console command arguments / options
+     *
+     * @return array
+     */
+    protected function getArguments()
+    {
+        return [
+            ['file', InputArgument::REQUIRED, 'Filepath of .sta containing the bank transactions'],
```
Comment (cschra) [inline modules/OverdueDebts/Console/ParseMt940.php]: where is voucherNr?

File: modules/OverdueDebts/Entities/TransactionParser.php [hunk 3]
```diff
@@ -39,13 +39,13 @@ private static function detectParser($text)
      *
      * @return object
      */
-    public function parse(\Kingsquare\Banking\Transaction $transaction)
+    public function parse(\Kingsquare\Banking\Transaction $transaction, $debt)
```
Comment (cschra) [inline modules/OverdueDebts/Entities/TransactionParser.php]: if we start typehinting, we should be consistent and typehint where it is possible ->
```php
use Modules\OverdueDebts\Entities\Debt;
use Kingsquare\Banking\Transaction;
```
```php
 public function parse(Transaction $transaction, Debt $debt)

File: modules/OverdueDebts/Entities/Mt940Parser.php [hunk 4]
```diff
@@ -113,4 +113,29 @@ public function parse($filepath, $voucherNr)
 
         // d($transactions, $statements, str_replace(':61:', "\r\n---------------\r\n:61:", $mt940));
     }
+
+    /**
+     * @param array
+     */
+    public static function addDebtButton($data)
```
Comment (cschra) [inline modules/OverdueDebts/Entities/Mt940Parser.php]: **this is no "model Code" -> this belongs into view!**

File: modules/OverdueDebts/Http/Controllers/DebtController.php [hunk 5]
```diff
@@ -88,4 +89,22 @@ public function result_datatables_ajax()
     {
         return parent::index_datatables_ajax();
     }
+
+    public function quickAdd()
+    {
+        // Replace Button in log with info that debt was added
+        $data = \Request::all();
+        unset($data['_untoken']);
```
Comment (cschra) [inline modules/OverdueDebts/Http/Controllers/DebtController.php]: what is _untoken?
Do you mean _token?

File: modules/OverdueDebts/Entities/DefaultTransactionParser.php [hunk 6]
```diff
@@ -334,6 +338,26 @@ private function addDebtBySpecialMatch($contract)
...
+            // Type cast to string so that data in addDebtButton() is encoded same way as
+            // done during ajax request when debt shall be added - see DebtController@quickAdd
+            'amount' => (string) $this->debt->amount,
+            'contract_id' => (string) $contract_id,
+            'date' => $this->debt->date,
+            'description' => $this->debt->description,
+            'voucher_nr' => $this->debt->voucher_nr,
+        ];
+
+        return Mt940Parser::addDebtButton($data);
```
Comment (cschra) [inline modules/OverdueDebts/Entities/DefaultTransactionParser.php]: use this to generate json and send it to view.

File: modules/OverdueDebts/Entities/Mt940Parser.php [hunk 7]
```diff
@@ -113,4 +113,29 @@ public function parse($filepath, $voucherNr)
...
+        ksort($data);
+
+        $html = '';
+        $html = '<button class="btn btn-link btn-xs addDebt" type="submit" value=\''.json_encode($data).'\'>'.
+        $html .= trans('overduedebts::view.debt.add').'</button>';
+
+        return $html;
+    }
+
+    public static function addedDebtInfo()
```
Comment (cschra) [inline modules/OverdueDebts/Entities/Mt940Parser.php]: this solution violates the MVC Principle. 
We have some code like this, but we should really try not to add more of it

File: modules/OverdueDebts/Http/Controllers/DebtController.php (hunk 5 above)
Comment (NinoRy) [reply modules/OverdueDebts/Http/Controllers/DebtController.php]: yes. fixed

File: modules/BillingBase/Http/Controllers/SettlementRunController.php (hunk 1 above)
Comment (NinoRy) [reply modules/BillingBase/Http/Controllers/SettlementRunController.php]: done

File: modules/OverdueDebts/Console/ParseMt940.php (hunk 2 above)
Comment (NinoRy) [reply modules/OverdueDebts/Console/ParseMt940.php]: added

File: modules/OverdueDebts/Entities/Mt940Parser.php (hunk 4 above)
Comment (NinoRy) [reply modules/OverdueDebts/Entities/Mt940Parser.php]: unfortunately there is no view and no controller here

File: modules/OverdueDebts/Entities/Mt940Parser.php (hunk 7 above)
Comment (cschra) [reply modules/OverdueDebts/Entities/Mt940Parser.php]: i created #782  to adress this

**Pull request merged**